src/framework/components/rigid-body/system.js

```diff
@@ -540,11 +550,15 @@ class RigidBodyComponentSystem extends ComponentSystem {
                     const result = new RaycastResult(
                         body.entity,
                         new Vec3(point.x(), point.y(), point.z()),
-                        new Vec3(normal.x(), normal.y(), normal.z())
+                        new Vec3(normal.x(), normal.y(), normal.z()),
+                        hitFractions.at(i)
                     );
+
                     results.push(result);
                 }
             }
+
+            results.sort((a, b) => a.hitFraction - b.hitFraction);
```

Comment:
Have you verified that `rayCallback.get_m_collisionObjects` is not already sorted?

Reply:
Yep, in fact it seems to be reverse-sorted but considering how many possibilities there are in a 3D space I can't 100% sure confirm it is reverse-sorted. All I can confirm is that it's not sorted. If you want to test it out on a basic app with `raycastAll(new pc.Vec3(0, 2, 2), new pc.Vec3(0, -2, -2))` you will hit box then plane while it goes through the box before the plane.